File: app/src/main/java/remix/myplayer/ui/adapter/ChildHolderAdapter.java
package remix.myplayer.ui.adapter;

import static remix.myplayer.helper.MusicServiceRemote.setPlayQueue;
import static remix.myplayer.request.ImageUriRequest.SMALL_IMAGE_SIZE;
import static remix.myplayer.theme.ThemeStore.getHighLightTextColor;
import static remix.myplayer.theme.ThemeStore.getTextColorPrimary;
import static remix.myplayer.util.ImageUriUtil.getSearchRequestWithAlbumType;
import static remix.myplayer.util.MusicUtil.makeCmdIntent;

import android.annotation.SuppressLint;
import android.content.Context;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.PopupMenu;
import androidx.recyclerview.widget.RecyclerView;
import android.text.TextUtils;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.TextView;
import butterknife.BindView;
import com.facebook.drawee.view.SimpleDraweeView;
import com.github.promeg.pinyinhelper.Pinyin;
import remix.myplayer.R;
import remix.myplayer.bean.mp3.Song;
import remix.myplayer.helper.MusicServiceRemote;
import remix.myplayer.misc.menu.SongPopupListener;
import remix.myplayer.service.Command;
import remix.myplayer.theme.Theme;
import remix.myplayer.theme.ThemeStore;
import remix.myplayer.ui.adapter.holder.BaseViewHolder;
import remix.myplayer.ui.misc.MultipleChoice;
import remix.myplayer.ui.widget.fastcroll_recyclerview.FastScroller;
import remix.myplayer.util.Constants;
import remix.myplayer.util.ToastUtil;

/**
 * Created by taeja on 16-6-24.
 */
@SuppressLint("RestrictedApi")
public class ChildHolderAdapter extends HeaderAdapter<Song, BaseViewHolder> implements
    FastScroller.SectionIndexer {

  private int mType;
  private String mArg;

  private Song mLastPlaySong = MusicServiceRemote.getCurrentSong();

  public ChildHolderAdapter(int layoutId, int type, String arg,
      MultipleChoice multiChoice, RecyclerView recyclerView) {
    super(layoutId, multiChoice, recyclerView);
    this.mType = type;
    this.mArg = arg;
  }

  @NonNull
  @Override
  public BaseViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
    return viewType == TYPE_HEADER ?
        new SongAdapter.HeaderHolder(
            LayoutInflater.from(parent.getContext())
                .inflate(R.layout.layout_header_1, parent, false)) :
        new ChildHolderViewHolder(
            LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_song_recycle, parent, false));
  }


  @Override
  public void onViewRecycled(@NonNull BaseViewHolder holder) {
    super.onViewRecycled(holder);
    disposeLoad(holder);
  }

  @Override
  protected void convert(final BaseViewHolder baseHolder, final Song song, int position) {
    final Context context = baseHolder.itemView.getContext();
    if (position == 0) {
      final SongAdapter.HeaderHolder headerHolder = (SongAdapter.HeaderHolder) baseHolder;
      //没有歌曲时隐藏
      if (mDatas == null || mDatas.size() == 0) {
        headerHolder.mRoot.setVisibility(View.GONE);
        return;
      }

      headerHolder.mIvShuffle.setImageDrawable(
          Theme.tintVectorDrawable(context, R.drawable.ic_shuffle_white_24dp,
              ThemeStore.getAccentColor())
      );
      headerHolder.mTvShuffle.setText(context.getString(R.string.play_random, getItemCount() - 1));

      //显示当前排序方式
      headerHolder.mRoot.setOnClickListener(v -> {
        //设置正在播放列表
        if (mDatas == null || mDatas.isEmpty()) {
          ToastUtil.show(context, R.string.no_song);
          return;
        }
        setPlayQueue(mDatas, makeCmdIntent(Command.NEXT, true));
      });
      return;
    }

    final ChildHolderViewHolder holder = (ChildHolderViewHolder) baseHolder;
    if (song == null || song.getId() < 0 || song.getTitle()
        .equals(context.getString(R.string.song_lose_effect))) {
      holder.mName.setText(R.string.song_lose_effect);
      holder.mButton.setVisibility(View.INVISIBLE);
    } else {
      holder.mButton.setVisibility(View.VISIBLE);

      //封面
      holder.mImage.setTag(
          setImage(holder.mImage, getSearchRequestWithAlbumType(song), SMALL_IMAGE_SIZE, position));

      //高亮
      if (MusicServiceRemote.getCurrentSong().getId() == song.getId()) {
        mLastPlaySong = song;
        holder.mName.setTextColor(getHighLightTextColor());
        holder.mIndicator.setVisibility(View.VISIBLE);
      } else {
        holder.mName.setTextColor(getTextColorPrimary());
        holder.mIndicator.setVisibility(View.GONE);
      }
      holder.mIndicator.setBackgroundColor(getHighLightTextColor());

      //设置标题
      holder.mName.setText(song.getShowName());

      //艺术家与专辑
      holder.mOther.setText(String.format("%s-%s", song.getArtist(), song.getAlbum()));

      if (holder.mButton != null) {
        //设置按钮着色
        int tintColor = ThemeStore.getLibraryBtnColor();
        Theme.tintDrawable(holder.mButton, R.drawable.icon_player_more, tintColor);

        holder.mButton.setOnClickListener(v -> {
          if (mChoice.isActive()) {
            return;
          }
          final PopupMenu popupMenu = new PopupMenu(context, holder.mButton, Gravity.END);
          popupMenu.getMenuInflater().inflate(R.menu.menu_song_item, popupMenu.getMenu());
          popupMenu.setOnMenuItemClickListener(
              new SongPopupListener((AppCompatActivity) context, song, mType == Constants.PLAYLIST,
                  mArg));
          popupMenu.show();
        });
      }
    }

    if (holder.mContainer != null && mOnItemClickListener != null) {
      holder.mContainer.setOnClickListener(v -> {
        if (holder.getAdapterPosition() - 1 < 0) {
          ToastUtil.show(context, R.string.illegal_arg);
          return;
        }
        if (song != null && song.getId() > 0) {
          mOnItemClickListener.onItemClick(v, holder.getAdapterPosition() - 1);
        }
      });
      holder.mContainer.setOnLongClickListener(v -> {
        if (holder.getAdapterPosition() - 1 < 0) {
          ToastUtil.show(context, R.string.illegal_arg);
          return true;
        }
        mOnItemClickListener.onItemLongClick(v, holder.getAdapterPosition() - 1);
        return true;
      });
    }

    holder.mContainer.setSelected(mChoice.isPositionCheck(position - 1));
  }

  @Override
  public String getSectionText(int position) {
    if (position == 0) {
      return "";
    }
    if (mDatas != null && mDatas.size() > 0 && position < mDatas.size()
        && mDatas.get(position - 1) != null) {
      String title = mDatas.get(position - 1).getTitle();
      return !TextUtils.isEmpty(title) ? (Pinyin.toPinyin(title.charAt(0))).toUpperCase()
          .substring(0, 1) : "";
    }
    return "";
  }

  public void updatePlayingSong() {
    final Song currentSong = MusicServiceRemote.getCurrentSong();
    if (currentSong.getId() == -1 || currentSong.getId() == mLastPlaySong.getId()) {
      return;
    }

    if (mDatas != null && mDatas.indexOf(currentSong) >= 0) {
      // 找到新的高亮歌曲
      final int index = mDatas.indexOf(currentSong) + 1;
      final int lastIndex = mDatas.indexOf(mLastPlaySong) + 1;

      ChildHolderViewHolder newHolder = null;
      if (mRecyclerView.findViewHolderForAdapterPosition(index) instanceof ChildHolderViewHolder) {
        newHolder = (ChildHolderViewHolder) mRecyclerView.findViewHolderForAdapterPosition(index);
      }
      ChildHolderViewHolder oldHolder = null;
      if (mRecyclerView
          .findViewHolderForAdapterPosition(lastIndex) instanceof ChildHolderViewHolder) {
        oldHolder = (ChildHolderViewHolder) mRecyclerView
            .findViewHolderForAdapterPosition(lastIndex);
      }

      if (newHolder != null) {
        newHolder.mName.setTextColor(getHighLightTextColor());
        newHolder.mIndicator.setVisibility(View.VISIBLE);
      }

      if (oldHolder != null) {
        oldHolder.mName.setTextColor(getTextColorPrimary());
        oldHolder.mIndicator.setVisibility(View.GONE);
      }
      mLastPlaySong = currentSong;
    }
  }

  static class ChildHolderViewHolder extends BaseViewHolder {

    @BindView(R.id.song_head_image)
    SimpleDraweeView mImage;
    @BindView(R.id.song_title)
    TextView mName;
    @BindView(R.id.song_other)
    TextView mOther;
    @BindView(R.id.song_button)
    public ImageButton mButton;
    public View mContainer;
    @BindView(R.id.indicator)
    View mIndicator;

    ChildHolderViewHolder(View itemView) {
      super(itemView);
      mContainer = itemView;
    }
  }
}

File: app/src/main/java/remix/myplayer/ui/adapter/ChildHolderAdapter.kt
package remix.myplayer.ui.adapter

import android.annotation.SuppressLint
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.PopupMenu
import androidx.recyclerview.widget.RecyclerView
import com.github.promeg.pinyinhelper.Pinyin
import remix.myplayer.App
import remix.myplayer.R
import remix.myplayer.bean.mp3.Song
import remix.myplayer.databinding.ItemSongRecycleBinding
import remix.myplayer.helper.MusicServiceRemote.getCurrentSong
import remix.myplayer.helper.MusicServiceRemote.setPlayQueue
import remix.myplayer.helper.SortOrder
import remix.myplayer.misc.menu.SongPopupListener
import remix.myplayer.request.ImageUriRequest
import remix.myplayer.service.Command
import remix.myplayer.theme.Theme
import remix.myplayer.theme.ThemeStore.accentColor
import remix.myplayer.theme.ThemeStore.highLightTextColor
import remix.myplayer.theme.ThemeStore.libraryBtnColor
import remix.myplayer.theme.ThemeStore.textColorPrimary
import remix.myplayer.ui.adapter.holder.BaseViewHolder
import remix.myplayer.ui.misc.MultipleChoice
import remix.myplayer.ui.widget.fastcroll_recyclerview.FastScroller
import remix.myplayer.util.Constants
import remix.myplayer.util.ImageUriUtil
import remix.myplayer.util.MusicUtil
import remix.myplayer.util.SPUtil
import remix.myplayer.util.SPUtil.SETTING_KEY
import remix.myplayer.util.ToastUtil
import java.util.*

/**
 * Created by taeja on 16-6-24.
 */
@SuppressLint("RestrictedApi")
open class ChildHolderAdapter(layoutId: Int, private val type: Int, private val arg: String, multiChoice: MultipleChoice<Song>, recyclerView: RecyclerView)
  : HeaderAdapter<Song, BaseViewHolder>(layoutId, multiChoice, recyclerView), FastScroller.SectionIndexer {

  private var lastPlaySong = getCurrentSong()

  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder {
    return if (viewType == TYPE_HEADER) SongAdapter.HeaderHolder(
        LayoutInflater.from(parent.context)
            .inflate(R.layout.layout_header_1, parent, false)) else ChildHolderViewHolder(
        ItemSongRecycleBinding.inflate(LayoutInflater.from(parent.context), parent,
            false))
  }

  override fun onViewRecycled(holder: BaseViewHolder) {
    super.onViewRecycled(holder)
    disposeLoad(holder)
  }

  override fun convert(holder: BaseViewHolder, data: Song?, position: Int) {
    val context = holder.itemView.context
    if (position == 0) {
      val headerHolder = holder as SongAdapter.HeaderHolder
      //没有歌曲时隐藏
      if (dataList.size == 0) {
        headerHolder.binding.root.visibility = View.GONE
        return
      }
      headerHolder.binding.root.visibility = View.VISIBLE
      headerHolder.binding.playShuffleButton.setImageDrawable(
          Theme.tintVectorDrawable(context, R.drawable.ic_shuffle_white_24dp,
              accentColor)
      )
      headerHolder.binding.tvShuffleCount.text = context.getString(R.string.play_random, itemCount - 1)

      //显示当前排序方式
      headerHolder.binding.root.setOnClickListener { v: View? ->
        //设置正在播放列表
        if (dataList.isEmpty()) {
          ToastUtil.show(context, R.string.no_song)
          return@setOnClickListener
        }
        setPlayQueue(dataList, MusicUtil.makeCmdIntent(Command.NEXT, true))
      }
      return
    }

    if (data == null) {
      return
    }
    val holder = holder as ChildHolderViewHolder
    if (data.id < 0 || (data.title == context.getString(R.string.song_lose_effect))) {
      holder.binding.songTitle.setText(R.string.song_lose_effect)
      holder.binding.songButton.visibility = View.INVISIBLE
    } else {
      holder.binding.songButton.visibility = View.VISIBLE

      //封面
      holder.binding.songHeadImage.tag = setImage(holder.binding.songHeadImage, ImageUriUtil.getSearchRequestWithAlbumType(data), ImageUriRequest.SMALL_IMAGE_SIZE, position)

      //高亮
      if (getCurrentSong().id == data.id) {
        lastPlaySong = data
        holder.binding.songTitle.setTextColor(highLightTextColor)
        holder.binding.indicator.visibility = View.VISIBLE
      } else {
        holder.binding.songTitle.setTextColor(textColorPrimary)
        holder.binding.indicator.visibility = View.GONE
      }
      holder.binding.indicator.setBackgroundColor(highLightTextColor)

      //设置标题
      holder.binding.songTitle.text = data.showName

      //艺术家与专辑
      holder.binding.songOther.text = String.format("%s-%s", data.artist, data.album)
      //设置按钮着色
      val tintColor = libraryBtnColor
      Theme.tintDrawable(holder.binding.songButton, R.drawable.icon_player_more, tintColor)
      holder.binding.songButton.setOnClickListener { v: View? ->
        if (choice.isActive) {
          return@setOnClickListener
        }
        val popupMenu = PopupMenu(context, holder.binding.songButton, Gravity.END)
        popupMenu.menuInflater.inflate(R.menu.menu_song_item, popupMenu.menu)
        popupMenu.setOnMenuItemClickListener(
            SongPopupListener((context as AppCompatActivity), data, type == Constants.PLAYLIST,
                arg))
        popupMenu.show()
      }
    }
    if (onItemClickListener != null) {
      holder.binding.root.setOnClickListener { v: View ->
        if (holder.adapterPosition - 1 < 0) {
          ToastUtil.show(context, R.string.illegal_arg)
          return@setOnClickListener
        }
        if (data.id > 0) {
          onItemClickListener?.onItemClick(v, holder.adapterPosition - 1)
        }
      }
      holder.binding.root.setOnLongClickListener { v: View ->
        if (holder.adapterPosition - 1 < 0) {
          ToastUtil.show(context, R.string.illegal_arg)
          return@setOnLongClickListener true
        }
        onItemClickListener?.onItemLongClick(v, holder.adapterPosition - 1)
        true
      }
    }
    holder.binding.root.isSelected = choice.isPositionCheck(position - 1)
  }

  override fun getSectionText(position: Int): String {
    if (position in 1..dataList.size) {
      val settingKey = when (type) {
        Constants.ALBUM -> SETTING_KEY.CHILD_ALBUM_SONG_SORT_ORDER
        Constants.ARTIST -> SETTING_KEY.CHILD_ARTIST_SONG_SORT_ORDER
        Constants.FOLDER -> SETTING_KEY.CHILD_FOLDER_SONG_SORT_ORDER
        Constants.PLAYLIST -> SETTING_KEY.CHILD_PLAYLIST_SONG_SORT_ORDER
        else -> null
      }
      if (settingKey != null) {
        val data = dataList[position - 1]
        val key = when (SPUtil.getValue(
          App.context,
          SETTING_KEY.NAME,
          settingKey,
          SortOrder.SONG_A_Z
        )) {
          SortOrder.SONG_A_Z, SortOrder.SONG_Z_A -> data.title
          SortOrder.ARTIST_A_Z, SortOrder.ARTIST_Z_A -> data.artist
          SortOrder.ALBUM_A_Z, SortOrder.ALBUM_Z_A -> data.album
          SortOrder.DISPLAY_NAME_A_Z, SortOrder.DISPLAY_NAME_Z_A -> data.displayName
          else -> ""
        }
        if (key.isNotEmpty())
          return Pinyin.toPinyin(key[0]).toUpperCase(Locale.getDefault()).substring(0, 1)
      }
    }
    return ""
  }

  fun updatePlayingSong() {
    val currentSong = getCurrentSong()
    if (currentSong.id == -1 || currentSong.id == lastPlaySong.id) {
      return
    }
    if (dataList.contains(currentSong)) {
      // 找到新的高亮歌曲
      val index = dataList.indexOf(currentSong) + 1
      val lastIndex = dataList.indexOf(lastPlaySong) + 1
      var newHolder: ChildHolderViewHolder? = null
      if (recyclerView.findViewHolderForAdapterPosition(index) is ChildHolderViewHolder) {
        newHolder = recyclerView.findViewHolderForAdapterPosition(index) as ChildHolderViewHolder?
      }
      var oldHolder: ChildHolderViewHolder? = null
      if (recyclerView
              .findViewHolderForAdapterPosition(lastIndex) is ChildHolderViewHolder) {
        oldHolder = recyclerView
            .findViewHolderForAdapterPosition(lastIndex) as ChildHolderViewHolder?
      }
      if (newHolder != null) {
        newHolder.binding.songTitle.setTextColor(highLightTextColor)
        newHolder.binding.indicator.visibility = View.VISIBLE
      }
      if (oldHolder != null) {
        oldHolder.binding.songTitle.setTextColor(textColorPrimary)
        oldHolder.binding.indicator.visibility = View.GONE
      }
      lastPlaySong = currentSong
    }
  }

  internal class ChildHolderViewHolder(val binding: ItemSongRecycleBinding) : BaseViewHolder(binding.root)
}